### ISSUE TITLE
Is it possible to use osde2e against pre-existing prod clusters?

### ISSUE THREAD

**Issue by josedev-union** [openshift/osde2e#669]
Hi team. Thank you for osde2e and your effort.
Seems likely the osde2e works only for clusters created by itself(with ocm).
Our expectation is to run the osde2e against the preexisting clusters to get the cluster status. Is there any functionality to do like this or any milestone?
We are willing to contribute for it.
Regards.
Jose.

**Comment by jeefy** [comment]
Hi Jose!

Currently, you could run this against an existing cluster given a valid Kubeconfig with cluster-admin.

My question in response: what kind of cluster status / info are you looking for? We might already collect/log what you're interested in.

Thanks!

-Jeff

**Comment by josedev-union** [comment]
Hi @jeefy. Thank you for your reply.
We are running openshift v4 clusters.
So the motivation is to run e2e test on the cluster as a cronjob so as to guarantee the healthy cluster continuously.
The current step is the identification step so not sure which metrics we need exactly.
We are planning to develop extensions/plugins to test our specific applications running on the cluster also.

Btw, can you share more information about the process to run the osde2e on the existing cluster?
And also the status/infor what we can get when run the test.
Thank you.
Jose

**Comment by josedev-union** [comment]
Hi @jeefy 
```
PROVIDER=mock TEST_KUBECONFIG=~/.kube/config ./osde2e test --configs prod
```
I am using this command to test the pre-existing cluster but get such an error.
```
addon timeout is 30
Will load config prod
Writing files to temporary directory /tmp/014839172
Outputting log to build log at /tmp/014839172/test_output.log
Found an existing Kubeconfig!
Running e2e tests...
Running Suite: OSD e2e suite
============================
Random Seed: 1611133594
Will run 179 of 179 specs

CLUSTER_ID set to 512ad644-5fab-4fde-b772-1690fc72b70e from OCM.
CLUSTER_NAME set to osde2e-tqknf from OCM.
CLUSTER_VERSION set to  from OCM.
CLOUD_PROVIDER_ID set to mock from OCM.
CLOUD_PROVIDER_REGION set to mock-region from OCM.
Found addons: 
Error while adding upgrade version property to cluster via OCM: AddProperty is unsupported by mock clusters
Waiting 135 minutes for cluster '512ad644-5fab-4fde-b772-1690fc72b70e' to be ready...
Error fetching cluster details from provider: couldn't find cluster in mock provider
Error fetching cluster details from provider: couldn't find cluster in mock provider
Error fetching cluster details from provider: couldn't find cluster in mock provider
Error fetching cluster details from provider: couldn't find cluster in mock provider
Error fetching cluster details from provider: couldn't find cluster in mock provider
```
Any configuration needed?

**Comment by mrbarge** [comment]
Hi @josedev-union !

Can you please try adding the `--skip-health-check` to your command-line arguments?

By default `osde2e` tries to verify the health of the cluster it installs and uses OCM for this purpose. The skip feature can (and should) be used for existing clusters where we can safely assume they are already healthy.

Another thing you may want to do is replace the `prod` argument to `--configs` with `e2e-suite`, which defines the range of tests that are tested. `prod` merely indicates to osde2e that it should use the Production OCM.  You can further refine what actual tests you want to run with the `--focus-tests` argument, which accepts a regex. eg `--focus-tests "Workload.*"`

So, in summary, try something like:

```
PROVIDER=mock TEST_KUBECONFIG==~/.kube/config ./osde2e test --skip-health-check --configs e2e-suite
```

Hope this helps,
Matt

**Comment by josedev-union** [comment]
HI @mrbarge 
yes, it works now. 
Seems the operator tests are hanging because the cluster doesn't have them. You mentioned about `--focus-tests` param to filter the tests.
https://github.com/openshift/osde2e#different-test-types
There are several kinds of tests but how can i specify the tests explicitly? And what is the test names?

And i have some more questions :)
Can i add some scheduling params to the resources which are created for the test? For example, i want to add `tolerations` or `nodeSelector` to the test pods. If it can be done by configuration, it would be helpful.

Thank you.
Jose

**Comment by mrbarge** [comment]
Glad to hear it worked!

Sorry I was not clearer with the `focus-tests` argument. The test names are those provided to `ginkgo.Describe()`, minus all the `[...]` tags.

For example, from [this test](https://github.com/openshift/osde2e/blob/661a8c9dae64a8297c47d2ee238e46bc64f02bcd/pkg/e2e/operators/configurealertmanager.go#L9):

```
var configureAlertManagerOperators string = "[Suite: operators] [OSD] Configure AlertManager Operator"
...
var _ = ginkgo.Describe(configureAlertManagerOperators, func() {
```

the test name is `Configure Alertmanager Operator` 

If you want to focus on multiple tests, separate the names by pipes, eg.

`--focus-tests "Configure Alertmanager Operator|RBAC Permissions Operator"`

You may alternatively find `--skip-tests` useful - it behaves the same as `--focus-tests` but instead is oriented towards skipping specific ones.

With regards to your latter question on configuring scheduling parameters, I don't think that is a supported feature right now (Jeefy may know more), but contributions are always welcome.

**Comment by josedev-union** [comment]
ok, I will.
seems `GINKGO_FOCUS` and `GINKGO_SKIP` are same configs.
The result of test is Failure :)
```
2021/01/20 14:41:11 osde2e-project runner is done
2021/01/20 14:41:13 Downloading osde2e-project.tgz
2021/01/20 14:41:13 Gathering Cluster State...
2021/01/20 14:41:40 Writing cluster state results
2021/01/20 14:41:40 Gathering cluster state from mock
2021/01/20 14:41:40 error getting Cluster state: couldn't find cluster in mock provider
2021/01/20 14:41:40 Addon cleanup: false
2021/01/20 14:41:40 Tests failed: please inspect logs for more details
testing failed
```
I can get test result files in /tmp/xxxx dir.
Anyway, I have 2 questions
- The test resources are created in `osde2e-xxx` project and still remaining even after test end. How can i remove test resources after test automatically?
- The test result is fail. But not sure what is the root cause of fail? Is this error a root cause? 
`error getting Cluster state: couldn't find cluster in mock provider`

**Comment by mrbarge** [comment]
At the moment, I'm pretty sure there is no feature built-in to osde2e to automatically remove the test resources.

However, they may help you with your second question. You may find more information in the JUnit-formatted xml file that captures the test results. This file can be found in `/tmp/xxxx/artifacts/` directory, or possibly `/tmp/xxxx/artifacts/install/`. Here is one example of a failed Ginkgo test; the specific assertion that failed is noted as part of it:

```
<testcase name="[install] [Suite: operators] AlertmanagerInhibitions should exist" classname="OSD e2e suite" time="0.491891158">
<failure type="Failure">/go/src/github.com/openshift/osde2e/pkg/e2e/osd/inhibitions.go:35 ClusterOperatorDown inhibits ClusterOperatorDegraded Expected <bool>: false to equal <bool>: true /go/src/github.com/openshift/osde2e/pkg/e2e/osd/inhibitions.go:141</failure>
</testcase>
```

I think the `error getting Cluster state: couldn't find cluster in mock provider` error can be safely ignored, it's something that maybe we need to either hide or make less look like an error when the mock provider is used.

**Comment by josedev-union** [comment]
@mrbarge @jeefy 
yes, we can improve such things we are facing in mock provider.

Now let me clarify more things :)
- seems the basic question but What is the main goal and usecases of osde2e?
It includes conformance test, must-gathering, operator tests, etc. Ofc, we can filter the test cases but I feel like osde2e test if the specific ocp version is ok, if the specific operators are running ok on container env(more specifically on this ocp version), if the specific apps are running on real cluster.. etc.
For such tests, only test clusters are enough so once the test ends, we can destroy it.

As i mentioned already, we are trying to use this for existing production clusters to keep all services well.
We are about to develop such test project and feel osde2e is very similar. So want to sync the motivation and roadmap with you.

- Can we make test extensions as plugins?
Now seems to write a new test scenario, need to add it to the main repo. But any specific test cases can be dedicated for the organizations. So it is great if we make the tests as plugin(like vault auth/secret engines). Then we can maintain plugins as separate repos.

We want much wider collaboration. 
Thank you.
Jose.

**Comment by cuttingedge1109** [comment]
@mrbarge @jeefy Hope to hear from you soon on @josedev-union's questions.

**Comment by whereswaldon** [comment]
@josedev-union @cuttingedge1109 I hadn't seen this issue until now (I'm new). The purpose of `osde2e` is to run end to end tests against OpenShift Dedicated. It sounds like you're not using OpenShift Dedicated clusters, so many of the assumptions and design constraints that shaped `osde2e` may not hold for you.

Making the tests more pluggable is certainly possible. We already support a form of this for addon testing. Addon authors can provide an OCI container image that runs their tests and stores the results at a well-known path in the container's filesystem. OSDe2e can then extract those results and summarize them alongside the builtin test results.
